### PR TITLE
Core: Implement SoFCTransform node

### PR DESCRIPTION
**Pull request by wwmayer**
In its doAction() method it only sets the model matrix if it's not the identity matrix. This improves the issue #7606

### CODE REVIEW

**Comment by mwganson** [comment]
If skipping the mult() function where it isn't needed helps, then that implies some sort of stack space issue?

**Comment by hyarion** [comment]
Reading through the source code of coin, it looks to me like the only difference this would do is to avoid removing the flags `FLG_IDENTITY` and `FLG_COMBINED` in the class `SoModelMatrixElement` when trying to multiply with an identity matrix. 
Numerically, coin's `SoTransform` node should still use an identity matrix for calculation as the `setTransform` function doesn't do anything when identity rotations, null scaling, and zero-vector is used. And multiplying with an identity matrix should get the original matrix as 1 and 0 are exact values in ieee 754. (I could have missed something as I've only read through the code, not actually tested it).

So if this change actually fixes the issue, which I assume it does since you created the PR, it looks to me like a bug which should be fixed upstream instead of here. I do understand that this might take time and a workaround would be needed in the meantime.

Do you know what the issue in coin is, which makes it fail to handle a multiplication with an identity matrix?

As this is a workaround for a coin issue (presumably unknown as no reference is provided). Would it be possible to:
1. Create a test case which shows the issue in coin, so we can be report it upstream?
2. Add the test case to FC so we know when it is solved upstream so we can remove the workaround here?

**Comment by wwmayer** [comment]
> Create a test case which shows the issue in coin, so we can be report it upstream?

`https://github.com/coin3d/coin/issues/534`

> Add the test case to FC so we know when it is solved upstream so we can remove the workaround here?

This isn't (easily) possible because how would you test that not everything is rendered. The strange thing is that when hovering over the not rendered part the ray pick action still finds the geometry.

> Do you know what the issue in coin is, which makes it fail to handle a multiplication with an identity matrix?

No. I guess it's related to the used SoShape node. For SoCube it works but for SoIndexedFaceSet it fails.

**Comment by mwganson** [comment]
Perhaps a screen image could be captured as part of a test, then a QPixmap could be inspected.

**Comment by wwmayer** [comment]
Well, it's confirmed from Coin devs that it's a bug there and as soon as it's fixed we get a notification.

**Comment by hyarion** [comment]
I just want to avoid a situation where workarounds/hacks are left in the code base and people start to wonder why it was added and if it is safe to remove or not. Adding a single line comment might be enough, like:

> SoFCTransform only exists as a workaround for https://github.com/coin3d/coin/issues/534, when that has been fixed upstream then this file should be removed and all instances should go back to use SoTransform again.

Would that be worth considering?

**Comment by mwganson** [comment]
I wonder if this PR speeds up rendering?  It's a check for the identity matrix, which I would think is a very quick check, compared to calling an unnecessary (not to mention buggy) function that might have a lot of overhead.

**Comment by adrianinsaval** [comment]
> > SoFCTransform only exists as a workaround for https://github.com/coin3d/coin/issues/534, when that has been fixed upstream then this file should be removed and all instances should go back to use SoTransform again.

I would say "fixed in all supported versions of coin"

File: src/Gui/Inventor/SoFCTransform.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+#ifndef GUI_INVENTOR_SOFCTRANSFORM_H
+#define GUI_INVENTOR_SOFCTRANSFORM_H
+
+#include <Inventor/nodes/SoTransform.h>
+#include <FCGlobal.h>
+
+namespace Gui
+{
+
+class GuiExport SoFCTransform : public SoTransform
```
Comment (hyarion) [inline src/Gui/Inventor/SoFCTransform.h:33]:
```suggestion
/**
 * @class SoFCTransform
 * @brief A temporary workaround for coin3d/coin#534.
 * 
 * This class is implemented as a workaround for a specific bug in Coin3D.
 * Refer to the issue tracker: https://github.com/coin3d/coin/issues/534
 * Once resolved in all supported versions of Coin3D, this class should be
 * removed and all instances should revert to using SoTransform.
 */
class GuiExport SoFCTransform : public SoTransform
```

File: src/Gui/ViewProvider.cpp [hunk 2]
```diff
@@ -104,7 +105,7 @@ ViewProvider::ViewProvider()
     pcRoot->ref();
     pcModeSwitch = new SoSwitch();
     pcModeSwitch->ref();
-    pcTransform  = new SoTransform();
+    pcTransform  = new SoFCTransform();
```
Comment (hyarion) [inline src/Gui/ViewProvider.cpp:108]:
```suggestion
    // Using SoFCTransform as a workaround for coin3d/coin#534.
    // When fixed, replace with SoTransform.
    pcTransform  = new SoFCTransform();
```

**Review by hyarion — COMMENTED** [review]
Here's a suggestion how the information could be added.

(I changed wording and included the suggestion from adrianinsaval)

File: src/Gui/ViewProvider.cpp (hunk 2 above)
Comment (wwmayer) [reply src/Gui/ViewProvider.cpp:108]: I don't think we need this extra comment here. Once SoFCTransform should be changed back to SoTransform the compiler will tell us anyway all locations.

File: src/Gui/Inventor/SoFCTransform.h (hunk 1 above)
Comment (hyarion) [reply src/Gui/Inventor/SoFCTransform.h:33]: @wwmayer ok, let's try again then :)
```suggestion
/**
 * @class SoFCTransform
 * @brief A temporary workaround for coin3d/coin#534.
 * 
 * This class is a workaround for a missing feature to reduce the OpenGL stack size.
 * The issue was reported here: https://github.com/coin3d/coin/issues/534
 * And was merged here: https://github.com/coin3d/coin/pull/535
 * 
 * Once this feature is available in all supported versions of Coin3D, this class should
 * be removed and all instances should revert to using SoTransform.
 */
class GuiExport SoFCTransform : public SoTransform
```

**Comment by maxwxyz** [comment]
PR got some conflicts because of the long queue of open PRs

**Comment by adrianinsaval** [comment]
@hyarion can you check if this backportable given that there was a conflict at some pouint

**Comment by adrianinsaval** [comment]
Backported 8de5d74

**Pull request merged**